### PR TITLE
greenhills support: fix the build warning for ghs platform

### PR DESCRIPTION
**Pull request by extinguish**
## Summary
1. fix the enum type mixed using warning
2. fix the undefined macro using warning
3. fix the unknown attribute handling warning
4. fix the declaration out of visible warning

### CODE REVIEW

File: include/nuttx/sched.h [hunk 1]
```diff
@@ -72,6 +72,18 @@
 #  define CONFIG_SCHED_SPORADIC_MAXREPL 3
 #endif
 
+#ifndef CONFIG_SCHED_CRITMONITOR_MAXTIME_THREAD
+#  define CONFIG_SCHED_CRITMONITOR_MAXTIME_THREAD -1
```
Comment (anchao) [inline include/nuttx/sched.h:76]: These symbols are only used in sched_critmonitor.c, why are there these errors? You need to align your code with the mainline:
https://github.com/apache/nuttx/blob/master/sched/sched/sched_critmonitor.c#L41-L51

File: include/nuttx/compiler.h [hunk 2]
```diff
@@ -254,11 +254,21 @@
 
 /* The nooptimiziation_function attribute no optimize */
 
-#  define nooptimiziation_function __attribute__((optimize(0)))
+#  if defined(__clang__)
```
Comment (anchao) [inline include/nuttx/compiler.h:257]: could we separate the GHS chapters from the GNUC? They have a lot of syntax differences.

**Review by anchao — CHANGES_REQUESTED** [review]

**Comment by extinguish** [comment]
@anchao the review comments are fixed, please help to make a review, very grateflu thanks ^_^

File: include/nuttx/compiler.h [hunk 3]
```diff
@@ -1110,6 +1110,18 @@
 
 #endif
 
+/* Green Hills Software definitions *****************************************/
+
+#if defined(__ghs__)
```
Comment (pkarashchenko) [inline include/nuttx/compiler.h:1115]: I think this should be placed under `#ifdef __GNUC__` block above.

Comment (pkarashchenko) [reply include/nuttx/compiler.h:1115]: Sorry, I missed https://github.com/apache/nuttx/pull/12886#discussion_r1713097499 when was posting a review.
Still I think that if greenhills compiler  defines `__GNUC__` the `#if defined(__ghs__)` should be encapsulated under `__GNUC__ ` branch and not at the end off all compilers selection

**Review by xiaoxiang781216 — APPROVED** [review]

File: include/nuttx/compiler.h (hunk 3 above)
Comment (xiaoxiang781216) [reply include/nuttx/compiler.h:1115]: ok

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**